### PR TITLE
added async support to colbert and lanchain.colbert libraries

### PR DESCRIPTION
**Pull request by epinzur**

### CODE REVIEW

File: libs/colbert/ragstack_colbert/cassandra_database.py [hunk 1]
```diff
@@ -27,14 +30,11 @@ class CassandraDatabase(BaseDatabase):
     """
 
     _table: ClusteredMetadataVectorCassandraTable
-    _instance = None
```
Comment (epinzur) [inline libs/colbert/ragstack_colbert/cassandra_database.py:30]: singleton class here was causing issues.

File: libs/langchain/ragstack_langchain/colbert/colbert_retriever.py [hunk 2]
```diff
@@ -37,9 +37,9 @@ def __init__(
         **kwargs: Any,
     ):
         super().__init__(retriever=retriever, k=k, **kwargs)
-        self._retriever = retriever
-        self._k = k
-        self._query_maxlen = query_maxlen
+        self.retriever = retriever
+        self.k = k
+        self.query_maxlen = query_maxlen
```
Comment (epinzur) [inline libs/langchain/ragstack_langchain/colbert/colbert_retriever.py:42]: fixes a bug instantiating the langchain colbert retriever from an existing vector store.

File: libs/colbert/ragstack_colbert/cassandra_database.py [hunk 3]
```diff
@@ -115,55 +130,158 @@ def add_chunks(self, chunks: List[Chunk]) -> List[Tuple[str, int]]:
...
-                        logging.error(
-                            f"issue inserting document data: {doc_id} chunk: {chunk_id}: {e}"
-                        )
-                    else:
-                        logging.error(
-                            f"issue inserting document embedding: {doc_id} chunk: {chunk_id} embedding: {embedding_id}: {e}"
-                        )
+
+        delay = 0.05  # Start with a 50 ms delay
+        while len(all_tasks) > 0:
```
Comment (nicoloboschi) [inline libs/colbert/ragstack_colbert/cassandra_database.py:264]: shall we raise an error if one or more tasks failed ?

Comment (epinzur) [reply libs/colbert/ragstack_colbert/cassandra_database.py:264]: I'll add that

**Comment by epinzur** [comment]
have found some issues when testing this on larger datasets.  do not merge until resolved.

**Comment by epinzur** [comment]
i think this is working well now.  I've also added exception raising if add or delete fails. 

@nicoloboschi can you pls re-review?

**Review by nicoloboschi — APPROVED** [review]
LGTM

File: libs/colbert/ragstack_colbert/base_vector_store.py [hunk 4]
```diff
@@ -76,7 +76,60 @@ def delete_chunks(self, doc_ids: List[str]) -> bool:
...
+
+        Parameters:
+            chunks (List[Chunk]): A list of `Chunk` instances to be stored.
+            concurrent_inserts (Optional[int]): How many concurrent inserts to make to the database. Defaults to 100.
+
+        Returns:
+            a list of tuples: (doc_id, chunk_id)
+        """
+
+    # handles LangChain add
```
Comment (cbornet) [inline libs/colbert/ragstack_colbert/base_vector_store.py:97]:
```suggestion
```

File: libs/colbert/ragstack_colbert/base_vector_store.py [hunk 5]
```diff
@@ -76,7 +76,60 @@ def delete_chunks(self, doc_ids: List[str]) -> bool:
...
+                                                   If provided, these are set 1 to 1 with the texts list.
+            doc_id (Optional[str]): The document id associated with the texts. If not provided,
+                                    it is generated.
+            concurrent_inserts (Optional[int]): How many concurrent inserts to make to the database. Defaults to 100.
+
+        Returns:
+            a list of tuples: (doc_id, chunk_id)
+        """
+
+    # handles LangChain and LlamaIndex delete
```
Comment (cbornet) [inline libs/colbert/ragstack_colbert/base_vector_store.py:121]:
```suggestion
```

**Pull request merged**